packages/app-utils/src/components/ConnectionBootstrap.tsx

```diff
@@ -56,47 +59,63 @@ export function ConnectionBootstrap({
     [api, client]
   );
 
-  useEffect(
-    function listenForShutdown() {
-      if (connection == null) return;
+  useEffect(() => {
+    if (connection == null) return;
 
-      function handleShutdown(event: CustomEvent): void {
-        const { detail } = event;
-        log.info('Shutdown', `${JSON.stringify(detail)}`);
-        setError(`Server shutdown: ${detail ?? 'Unknown reason'}`);
-      }
+    // handles the shutdown event
+    function handleShutdown(event: CustomEvent): void {
+      const { detail } = event;
+      log.info('Shutdown', `${JSON.stringify(detail)}`);
+      setError(`Server shutdown: ${detail ?? 'Unknown reason'}`);
+    }
+    const removerFn = connection.addEventListener(
+      api.IdeConnection.EVENT_SHUTDOWN,
+      handleShutdown
+    );
 
-      const removerFn = connection.addEventListener(
-        api.IdeConnection.EVENT_SHUTDOWN,
-        handleShutdown
+    // handles the auth failed event
+    function handleAuthFailed(event: CustomEvent): void {
+      const { detail } = event;
+      log.warn('Reconnect authentication failed', `${JSON.stringify(detail)}`);
+      setError(
+        `Reconnect authentication failed: ${detail ?? 'Unknown reason'}`
       );
-      return removerFn;
-    },
-    [api, connection]
-  );
+      setIsAuthFailedState(true);
+    }
+    const authFailed = connection.addEventListener(
+      dh.CoreClient.EVENT_RECONNECT_AUTH_FAILED,
+      handleAuthFailed
+    );
+
+    return authFailedState ? authFailed : removerFn;
+  }, [api, connection, authFailedState]);
 
   const objectFetcher = useCallback(
-    async (descriptor: dh.ide.VariableDescriptor) => {
+    async (descriptor: DhType.ide.VariableDescriptor) => {
       assertNotNull(connection, 'No connection available to fetch object with');
       return connection.getObject(sanitizeVariableDescriptor(descriptor));
     },
     [connection]
   );
 
-  if (connection == null || error != null) {
-    return (
-      <LoadingOverlay
-        data-testid="connection-bootstrap-loading"
-        isLoading={connection == null}
-        errorMessage={error != null ? `${error}` : undefined}
-      />
-    );
+  function handleRefresh(): void {
+    log.info('Refreshing application');
+    window.location.reload();
   }
 
   return (
-    <ConnectionContext.Provider value={connection}>
+    <ConnectionContext.Provider value={connection ?? null}>
       <ObjectFetcherContext.Provider value={objectFetcher}>
-        {children}
+        <>
+          {children}
+          <BasicModal
+            confirmButtonText="Refresh"
+            onConfirm={handleRefresh}
+            isOpen={authFailedState}
+            headerText="Authentication failed"
+            bodyText="Credentials are invalid. Please refresh your browser to try and reconnect."
+          />
+        </>
```

Comment:
The logic handling all of this should be removed from `AppMainContainer` now so it's not duplicated.